src/identity/caclient.rs

```diff
@@ -93,3 +125,97 @@ impl CertificateProvider for CaClient {
         Ok(certs)
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use std::time::Duration;
+
+    use super::{CaClient, CertificateProvider, IstioCertificateService};
+    use crate::{
+        identity::{Error, Identity},
+        tls,
+        xds::istio::ca::{IstioCertificateRequest, IstioCertificateResponse},
+    };
+    use async_trait::async_trait;
+
+    #[derive(Clone, Default)]
+    struct MockCertService {
+        response: Option<IstioCertificateResponse>,
+        error: Option<tonic::Status>,
+    }
+
+    impl MockCertService {
+        fn set_response(&mut self, res: IstioCertificateResponse) {
+            self.error = None;
+            self.response = Some(res);
+        }
+
+        #[allow(dead_code)]
+        fn set_error(&mut self, err: tonic::Status) {
+            self.response = None;
+            self.error = Some(err);
+        }
+    }
+
+    #[async_trait]
+    impl IstioCertificateService for MockCertService {
+        async fn create_certificate(
+            &mut self,
+            _request: IstioCertificateRequest,
+        ) -> Result<tonic::Response<IstioCertificateResponse>, tonic::Status> {
+            if let Some(res) = &self.response {
+                Ok(tonic::Response::new(res.clone()))
+            } else if let Some(err) = &self.error {
+                Err(err.clone())
+            } else {
+                Err(tonic::Status::not_found("response not setup yet"))
+            }
+        }
+    }
+
+    async fn test_ca_client_with_response(
+        res: IstioCertificateResponse,
+    ) -> Result<crate::tls::Certs, crate::identity::Error> {
+        let mut mock_service = Box::new(MockCertService::default());
+        mock_service.set_response(res);
```

Comment:
we will eventually need to be able to retain the underlying mock_service probably so we can do something like `set err; fetch; set ok` or the opposite to test refreshes. but we can refactor as needed when we get to that

Reply:
yeah just need RefCell on the inner part. But again, maybe that lib is worth it if we have more things to mock.